runtime/parser/expression.go

```diff
@@ -640,6 +640,12 @@ func defineLessThanOrTypeArgumentsExpression() {
 
 			} else {
 
+				// if the error specifically occurred during parsing restricted types,
+				// this is still an invocation and we should report that error instead
+				if strings.Contains(err.Error(), "restricted type") {
```

Comment:
Using the error message to detect this case is brittle: if the error message is changed, the code here will still compile, but will fail at run-time.

Instead of returning a generic `SyntaxError` for the restricted type, introduce and report a new error type, and check for it here.